File: SleepingRoutineForZy/Core/Models/StreakCalculator.swift
import Foundation

enum StreakCalculator {
    /// Night key = local calendar date of `startedAt`.
    /// Counts nights with non-nil `completedAt`.
    /// Anchor is today if now is at/after preferred bedtime, else yesterday.
    static func currentStreak(
        sessions: [SleepSessionRecord],
        bedtimeHour: Int,
        bedtimeMinute: Int,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> Int {
        var completedNights = Set<DateComponents>()
        for session in sessions where session.completedAt != nil {
            let parts = calendar.dateComponents([.year, .month, .day], from: session.startedAt)
            completedNights.insert(parts)
        }
        guard !completedNights.isEmpty else { return 0 }

        let hour = calendar.component(.hour, from: now)
        let minute = calendar.component(.minute, from: now)
        let afterBedtime = hour > bedtimeHour || (hour == bedtimeHour && minute >= bedtimeMinute)

        var cursor = calendar.startOfDay(for: now)
        if !afterBedtime {
            cursor = calendar.date(byAdding: .day, value: -1, to: cursor) ?? cursor
        }

        func key(_ date: Date) -> DateComponents {
            calendar.dateComponents([.year, .month, .day], from: date)
        }

        let previous = calendar.date(byAdding: .day, value: -1, to: cursor) ?? cursor
        if !completedNights.contains(key(cursor)) && !completedNights.contains(key(previous)) {
            return 0
        }
        if !completedNights.contains(key(cursor)) {
            cursor = previous
        }

        var streak = 0
        while completedNights.contains(key(cursor)) {
            streak += 1
            guard let next = calendar.date(byAdding: .day, value: -1, to: cursor) else { break }
            cursor = next
        }
        return streak
    }
}

File: flutter_app/lib/core/services/streak_calculator.dart
import '../models/models.dart';

/// Computes consecutive-night streaks from completed sleep sessions.
class StreakCalculator {
  /// Night key = local calendar date of [SleepSessionRecord.startedAt].
  /// A night counts when [completedAt] is non-null.
  /// Anchor night is today if [now] is at/after preferred bedtime, else yesterday.
  static int currentStreak({
    required List<SleepSessionRecord> sessions,
    required int bedtimeHour,
    required int bedtimeMinute,
    DateTime? now,
  }) {
    final clock = now ?? DateTime.now();
    final completedNights = <DateTime>{};
    for (final s in sessions) {
      if (s.completedAt == null) continue;
      final d = s.startedAt;
      completedNights.add(DateTime(d.year, d.month, d.day));
    }
    if (completedNights.isEmpty) return 0;

    final bedtimeToday = DateTime(
      clock.year,
      clock.month,
      clock.day,
      bedtimeHour,
      bedtimeMinute,
    );
    var cursor = !clock.isBefore(bedtimeToday)
        ? DateTime(clock.year, clock.month, clock.day)
        : DateTime(clock.year, clock.month, clock.day).subtract(const Duration(days: 1));

    // If neither anchor nor previous night has a session, streak is 0.
    if (!completedNights.contains(cursor) &&
        !completedNights.contains(cursor.subtract(const Duration(days: 1)))) {
      return 0;
    }
    if (!completedNights.contains(cursor)) {
      cursor = cursor.subtract(const Duration(days: 1));
    }

    var streak = 0;
    while (completedNights.contains(cursor)) {
      streak += 1;
      cursor = cursor.subtract(const Duration(days: 1));
    }
    return streak;
  }
}
